checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -33,6 +52,150 @@ public UpperBoundVisitor(BaseTypeChecker checker) {
     }
 
     /**
+     * This visits all local variable declarations and issues a warning if a dependent annotation is
+     * written on a local variable. This is necessary for the soundness of the reassignment code,
+     * which must unrefine all qualifiers, but which cannot collect all in-scope local variables. So
+     * that there are no local variables with qualifiers that are not in the store, this method
+     * forbids programmers from writing such qualifiers. This may introduce false positives if the
+     * programmer would like to write a non-primary annotation on a local variable. Note that final
+     * or effectively final are always permitted in dependent annotations, even in local variables,
+     * and will not trigger this warning (as no side-effect can effect them).
+     *
+     * <p>Also issues warnings if the expression in a dependent annotation is not permitted.
+     * Permitted expressions in dependent annotations must be at least one of the following:
+     *
+     * <ul>
+     *   <li>final or effectively final variables
+     *   <li>local variables
+     *   <li>private fields
+     *   <li>pure method calls all of whose arguments (including the receiver expression) are
+     *       composed only of expressions in this list
+     *   <li>accesses of a public final field whose access expression (sometimes called the
+     *       receiver) is composed only of expressions in this list
+     * </ul>
+     *
+     * <p>Any other expression results in a warning.
+     */
+    @Override
+    public Void visitVariable(VariableTree node, Void p) {
+        AnnotatedTypeMirror atm = atypeFactory.getAnnotatedTypeLhs(node);
+        AnnotationMirror anm = atm.getAnnotationInHierarchy(atypeFactory.UNKNOWN);
+        if (anm != null && AnnotationUtils.hasElementValue(anm, "value")) {
+            // This is a dependent annotation. If this is a local variable,
+            // issue a warning; dependent annotations should not be written on local variables.
+            Element elt = TreeUtils.elementFromDeclaration(node);
+            boolean allVariablesFinal = true;
+
+            UBQualifier qual = UBQualifier.createUBQualifier(anm);
+            if (qual.isLessThanLengthQualifier()) {
+                LessThanLengthOf ltl = (LessThanLengthOf) qual;
+                for (String array : ltl.getSequences()) {
+                    Receiver rec = getReceiverForCheckingDependentTypes(array, node);
+                    // covers final and effectively final variables
+                    boolean isFinal =
```

Comment:
All LocalVariable are not final, so this variable's name isn't correct.  (Use org.checkerframework.javacutil.ElementUtils#isEffectivelyFinal for local variables.)